Upgrade Ruby to 3.1

matthewford opened the pull request:
Ruby 2.6 is no longer receiving support and is end-of-life, and Ruby 2.7 will also be in March. Ruby 2.6 tests were removed from CI over a year ago (https://github.com/mastodon/mastodon/pull/16861)

This PR adds support for Ruby 3.1

I also documented and cleaned up the Gemfile.

This PR upgrades the CircleCI config and dev Ruby in Vagrant to 3.1. It also might be easier to review by each commit and in a split view.

Upgrades the Dockerfile to use Ruby 3.1

It also configures bootsnap to ignore node_modules.

To get the app running on 3.1 I needed to:

- [x] [add net-smtp](https://github.com/mastodon/mastodon/commit/e466998734b778a2d2cb39eb2acb720e469e24c9); this can be removed once Rails is upgraded to v7 (assuming the fix will be backported by then in the [mail](https://github.com/mikel/mail/pull/1472) gem)
- [x] Use a patched version of rails-settings-cached to work with psych 4
- [x] Use latest commit of of kt-paperclip for psych 4.

matthewford committed:
Add 3.1 to the build matrix

matthewford committed:
Update and Document Gemfile

To make it easier to see what gems are for, I have documents gems where the names are not clear of what they are used for.

I have also grouped them by use to make it easier to see similar gems.

Also discovered on newer ruby versions posix-spawn is not required.

Add  gem 'net-smtp' as it is not part of Ruby 3.1

Lock psych to version 3

Update Gemfile.lock

Allow to use Ruby 3.1

matthewford committed:
Update .gitignore

Add asdf tools version to gitignore

matthewford committed:
Upgrade ruby circleci orb to 2.0.0

update default ruby version to 3.1

matthewford committed:
Update .ruby-version

Use Ruby 3.1 in the Vargant file

matthewford committed:
Configure bootsnap to ignore node_modules

This allows us to skip non-ruby file like node_modules which slow down
bootsnap.

matthewford commented:
https://github.com/mastodon/mastodon/issues/20333

zunda commented on Gemfile.lock:
```diff
@@ -476,6 +480,7 @@ GEM
       pry (>= 0.13, < 0.15)
     pry-rails (0.3.9)
       pry (>= 0.10.4)
+    psych (3.3.4)
```
I'm not sure if locking the version of Psych is a productive option. Might we want to seek for modifying the code to be able to run on older and newer versions of Psych?

matthewford replied:
Would require I belive a Rails 7 upgrade

zunda replied:
[rails-settings-cached-2.8.2 is compatible with Rails 5.0.0 and up](https://github.com/huacnlee/rails-settings-cached/blob/v2.8.2/rails-settings-cached.gemspec#L28) and I guess we'd need to someday be able to use Psych >= 4 for safer code base.

On the other hand, it seems [rails-settings-cached-2 introduced a number of breaking changes](https://github.com/huacnlee/rails-settings-cached/releases/tag/v2.0.0) and there seems to be a long way for us to migrate from `config/settings.yml` to `app/models/setting.rb` and retain scoped settings.

Might there be expert who has more insight into this?

matthewford replied:
Upgrading [rails-settings-cached-2](https://github.com/huacnlee/rails-settings-cached/releases/tag/v2.0.0) would help, but I think i also saw a failure in Rails 6

matthewford commented:
> Any reason not to upgrade to Ruby 3.0 instead? It seems most (even rolling-release) distros don't have 3.1 available yet, and even 3.0 is only in the faster distros - https://pkgs.org/search/?q=ruby

3.0 is already supported in the CI build matrix.

matthewford commented:
The docker file definitely needs a review - its 2 am here, so I might look at it again in the morning.

ineffyble commented:
To bump the version in Docker you'll first need to modify the separate repo that builds Ruby for Docker. That's currently https://github.com/moritzheiber/ruby-jemalloc-docker because @moritzheiber and @Gargron are in the process of moving it to the mastodon org.

ineffyble commented on Dockerfile:
```diff
@@ -1,81 +1,90 @@
 
-FROM ghcr.io/moritzheiber/ruby-jemalloc:3.0.4-slim as ruby
-FROM node:${NODE_VERSION} as build
+FROM ruby:3.1.2-slim-bullseye as build
```
This won't work. The official ruby images don't have jemalloc, which is why the current version pulls a custom image.

matthewford replied:
@ineffyble While they do not have it, you can install and then enable it:

https://github.com/mastodon/mastodon/pull/21061/commits/204ecbe84e0e96c6e651042eeb7642bb8d6e4f63#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R68

https://github.com/mastodon/mastodon/pull/21061/commits/204ecbe84e0e96c6e651042eeb7642bb8d6e4f63#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R87

matthewford commented:
> To bump the version in Docker you'll first need to modify the separate repo that builds Ruby for Docker. That's currently https://github.com/moritzheiber/ruby-jemalloc-docker because @moritzheiber and @Gargron are in the process of moving it to the mastodon org.

Ah I was wondering if it was going to be moved to mastadon's org, in that case, I can remove the changes to the docker file, once it has been moved, we can then bump the repo, and then change it here.

moritzheiber commented:
I can easily add a version for 3.1.x alongside the current one, for testing future releases. Just let me know which official images you want supported or submit a PR to the repository 😃

matthewford commented:
@moritzheiber Thanks, removing the Dockerfile changes from this PR, it can be upgraded in another PR once the Dockerfile is updated

moritzheiber commented:
This is available now: https://github.com/moritzheiber/ruby-jemalloc-docker/pkgs/container/ruby-jemalloc/51641954?tag=3.1.2-slim

matthewford committed:
Update Dockerfile to use Ruby 3.1

matthewford commented:
@ineffyble rebased to resolve the gem conflict with main, and ran the i1n8 task that was failing

matthewford commented:
Also covers https://github.com/mastodon/mastodon/pull/21477/files

matthewford commented:
Also covers https://github.com/mastodon/mastodon/pull/21344

matthewford commented:
rebased against main, as boosnap was upgraded, this now just configures the new option to ignore node_modules

matthewford commented:
rebased again

maximecb commented:
@Gargron @ykzts this PR needs your approval 🙏 ❤️ 

Very interested in seeing this merged as Ruby 3.1 offers many performance improvements, and it will also allow us to experiment with running Mastodon with [YJIT](https://github.com/ruby/ruby/blob/master/doc/yjit/yjit.md) (included in Ruby 3.1).

zunda commented:
@Gargron might you have thoughts about whether we'll stick with rails-settings-cached at v0.6.6 or we'll want to migrate to a newer version? Because of the update of Psych from 3 to 4 on Ruby 3.1, rails-settings-cached-0.6.6 isn't compatible with Ruby >= 3.1.

It looks like we have a heavy dependency to rails-settings-cached-0.6.6 from files like `app/lib/settings/scoped_settings.rb`. I don't think I'd personally be able to migrate Mastodon to rails-settings-cached-2.8.2 which is compatible with wider range of Ruby versions >= 2.5.

This Pull Request works around the incompatibility of rails-settings-cached-0.6.6 to ruby-3.1 by installing  Psych 3 which is less secure than Psych 4. I'm actually against doing this as the whole code base becomes insecure than is indicated by the Ruby version.

If we're to be stuck with rails-settings-cached-0.6.6, I guess we'd want to rather make only this gem less secure. https://github.com/zunda/rails-settings-cached/tree/v0.6.6-aliases-true is what I have been doing to run Mastodon on Ruby 3.1 and 3.2 with a line like the below in `Gemfile`:

```
gem 'rails-settings-cached', '~> 0.6', git: 'https://github.com/zunda/rails-settings-cached.git', branch: 'v0.6.6-aliases-true'
```

matthewford commented:
That sounds like a good approach. I can bump Psyc and switch to your branch
till we upgrade the rails-settings-cached gem.

On Thu, 8 Dec 2022 at 00:06, zunda ***@***.***> wrote:

> @Gargron <https://github.com/Gargron> might you have thoughts about
> whether we'll stick with rails-settings-cached at v0.6.6 or we'll want to
> migrate to a newer version? Because of the update of Psych from 3 to 4 on
> Ruby 3.1, rails-settings-cached-0.6.6 isn't compatible with Ruby >= 3.1.
>
> It looks like we have a heavy dependency to rails-settings-cached-0.6.6
> from files like app/lib/settings/scoped_settings.rb. I don't think I'd
> personally be able to migrate Mastodon to rails-settings-cached-2.8.2 which
> is compatible with wider range of Ruby versions >= 2.5.
>
> This Pull Request works around the incompatibility of
> rails-settings-cached-0.6.6 to ruby-3.1 by installing Psych 3 which is less
> secure than Psych 4. I'm actually against doing this as the whole code base
> becomes insecure than is indicated by the Ruby version.
>
> If we're to be stuck with rails-settings-cached-0.6.6, I guess we'd want
> to rather make only this gem less secure.
> https://github.com/zunda/rails-settings-cached/tree/v0.6.6-aliases-true
> is what I have been doing to run Mastodon on Ruby 3.1 and 3.2 with a line
> like the below in Gemfile:
>
> gem 'rails-settings-cached', '~> 0.6', git: 'https://github.com/zunda/rails-settings-cached.git', branch: 'v0.6.6-aliases-true'
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mastodon/mastodon/pull/21061#issuecomment-1341779139>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAABV3HWRI2TSQPIEYY6HDWMEQ6XANCNFSM6AAAAAASFAF6XI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Sent from my iPhone

-- 
Bit Zesty Limited is a limited company registered in England and Wales. 
Registered number: 06883289. Registered office: 5 Colindeep Lane, London, 
England, NW9 6BX.

ineffyble commented:
@matthewford I managed to make CircleCI run, looks like some of the tests with Ruby 3.1 are failing

matthewford committed:
Use a branch of rails-settings-cached for Psych 4

https://github.com/mastodon/mastodon/pull/21061#issuecomment-1341779139

matthewford committed:
bump ci cache key

matthewford committed:
kt-paperclip with Psych 4 fix

Specs were failing due to paperclips' do_not_validate_attachment_file_type option not working properly

https://github.com/thoughtbot/paperclip/issues/2195

Paperclip: validate_media_type: false also broke the specs as it still expects a filename validation.

So adding in a validation to appease paperclips requirements.

matthewford commented:
> @matthewford I managed to make CircleCI run, looks like some of the tests with Ruby 3.1 are failing

I see; however, if I run the specs files individually or the entire suite locally they are passing 😑 

Some of these specs might be a bit flaky, I've rerun on my CircleCI, and I can see some failures, I will investigate

matthewford committed:
CircleCI missing imagemagic packages

Trying to fix convert issues:
  convert-im6.q16: unable to open image `{:output=': No such file or directory @ error/blob.c/OpenBlob/2924.
  convert-im6.q16: no decode delegate for this image format `' @ error/constitute.c/ReadImage/575.
  convert-im6.q16: unable to open image `q:a=': No such file or directory @ error/blob.c/OpenBlob/2924.
  convert-im6.q16: unable to open image `q:a=': No such file or directory @ error/blob.c/OpenBlob/2924.
  convert-im6.q16: no decode delegate for this image format `' @ error/constitute.c/ReadImage/575.
  ./lib/paperclip/lazy_thumbnail.rb:17:in `make'

ineffyble commented:
I suspect your changes have upgraded libyaml on CircleCI, meaning CI no longer has the load-bearing YAML indentation bug, which is why i18n is now failing. Oh dear.

matthewford commented:
Not sure why that's failing under 3.1 when I run the check locally it's fine:

```
~/D/G/mastodon ❯❯❯ bundle exec i18n-tasks check-normalized                                                                                                                        3.1 ✭
warning: parser/current is loading parser/ruby31, which recognizes3.1.2-compliant syntax, but you are running 3.1.0.
Please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
#StandWithUkraine
ruby -v
✓ Well done! All data is normalized
~/D/G/mastodon ❯❯❯ ruby -v                                                                                                                                                        3.1 ✭
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-darwin21]
```

ineffyble commented:
It's not about the Ruby version, it's about the version of the `libyaml` C library installed on your system. There's a bug that was fixed in the most recent version. CircleCI agents don't have that version usually, but I suspect that as part of the additional packages you've added, the new version is getting pulled in.

To replicate the issue locally you'll need to upgrade libyaml.

matthewford commented:
Is there a newer version than libyaml 0.2.5? That's what I've got installed

ineffyble commented:
Ah, the issue is the other direction then.

You need to downgrade your libyaml to the version with the bug, in order to match what CircleCI has. It's a real pain, happened to me too.

Alternatively you can just revert your changes to the locale files and push.

matthewford committed:
Run check-i18n on ubuntu-22.04

To match CI

matthewford commented:
Alright, thanks @ineffyble. I removed the commit that ran the i18n-tasks normalize task

matthewford commented:
will take alook at the conflicts tomorrow

matthewford commented:
@tenderlove 's PR supersedes this. The gemfile refactor had too many conflicts to ever get merged unless someone is willing to do it quickly.

Closing for now

eregon commented:
For reference, that's https://github.com/mastodon/mastodon/pull/22585 (and https://github.com/mastodon/mastodon/pull/22928)
